### ISSUE TITLE
Add InsertOnDuplicateSetStep::setAllToExcluded to auto-set all insert column list columns to their EXCLUDED value

### ISSUE THREAD

**Issue by xathien** [jOOQ/jOOQ#7912]
Very slightly related to #5214 and #3172.

It would be nice for jOOQ to reuse its algorithm for `UpdatableRecord.update` when using `onDuplicateKeyUpdate()`. For example, given the schema
```sql
  CREATE TABLE foo (
    id uuid NOT NULL PRIMARY KEY,
    bar VARCHAR NOT NULL
  );
```

and the simple case:
```java
    FooRecord foo1 = context.newRecord(FOO);
    foo1.setId(randomUUID());
    foo1.setBar("bar1");
    FooRecord foo2 = context.newRecord(FOO);
    foo2.setId(randomUUID());
    foo2.setBar("bar2");
    
    context.insertInto(FOO)
        .set(foo1)
        .newRecord()
        .set(foo2)
        .execute();
    
    foo1.setBar("bar3");
    foo2.setBar("bar4");
    
    Foo excluded = FOO.as("excluded");
    
    context.insertInto(FOO)
        .set(foo1)
        .newRecord()
        .set(foo2)
        .onDuplicateKeyUpdate()
        .set(FOO.BAR, excluded.BAR)
        .execute();
```

the requirement to be extremely verbose in the `.set()` command is more tedious than need be, especially when dealing with MySQL's `values(bar)` or Postgres' `excluded.bar`. We get _close_ to the ideal behavior when using `.onDuplicateKeyUpdate().set(foo1).execute()` -- it will determine which values to update and set them correctly... but only as long as there is exactly one record.

Possibilities:
1) Allow the command `.onDuplicateKeyUpdate().execute()` directly, which could use the table name specified (`.insertInto(FOO)`) to determine the needed updates.
2) Barring type-safety or API issues with the previous, allow `.onDuplicateKeyUpdate().set(FOO).execute()` to instruct jOOQ to use that table to determine the needed updates.

Bonus points: allow a list of `Record`s to be specified to the initial `.set` so I don't have to do this number in my insert/upsert methods 😅
```java
    Table<R> excluded = table.as("excluded");

    seq(recordList.subList(0, recordList.size() - 1))
        .foldLeft(
            context.insertInto(table),
            (insertQuery, record) -> insertQuery.set(record).newRecord())
        .set(recordList.get(recordList.size() - 1))
        .onDuplicateKeyUpdate()
        .set(upsertFields.stream()
            .collect(toMap(identity(), excluded::field)))
        .execute();
```

I hope this is clear and feasible! I'm interested in any discussion or thoughts you have.

**Comment by lukaseder** [comment]
Thanks for your suggestions. I will digest them later in detail, to see what can really be done. I feel that further "improvements" of the `ON DUPLICATE KEY UPDATE` clause might not be the best approach.

H2 has a neat vendor-specific `MERGE` statement, similar to the `UPSERT` statement offered by HANA:

- http://h2database.com/html/grammar.html#merge
- https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.03/en-US/20fc06a7751910149892c0d09be21a38.html

In most cases, we really want to pass the record only once, and let the database *Do The Right Thing™*.

**Comment by KoichiSenada** [comment]
@lukaseder , and what about PostgreSQL?
The `DSLContext.mergeInto()` is not supported by that dialect.

I need to upsert a number of records into a table, which has both primary keys and unique keys, due to which conflicts are rather possible.

**Comment by lukaseder** [comment]
@KoichiSenada Is your issue really the same as this one? If in doubt, feel free to open a new one...

**Comment by lukaseder** [comment]
I see, thanks @KoichiSenada. We'll definitely consider offering a better solution here at some point

**Comment by MikePryadko** [comment]
Any news?) A year and a half has passed...
Looks like "milestones" have no sense :neutral_face:

**Comment by lukaseder** [comment]
Thanks for your comment, @MikePryadko.

We use milestones not for planning, or as a promise, but for prioritisation, and as a reminder to look at things again to see if they can be implemented in the next release.

**Comment by lukaseder** [comment]
I've encountered a few cases elsewhere, where this would have been useful. In the meantime, there's `UpdatableRecord.merge()` for a similar, single-record use-case. It would be interesting to have new API for the DML statement, and retrofit `UpdatableRecord.merge()` to use that new API.

**Comment by lukaseder** [comment]
For the record, the original feature request by @xathien included a request for:

> Bonus points: allow a list of `Record`s to be specified to the initial `.set` so I don't have to do this number in my insert/upsert methods 😅

This has been implemented a while ago via `InsertValuesStep[N]::valuesOfRecords`, see: https://github.com/jOOQ/jOOQ/issues/6604

**Comment by lukaseder** [comment]
So, removing the idea of this being related with the `set(record)` call for the `INSERT` part, let's focus on the actual feature, which is to be able to auto-set all non-key columns in the `UPDATE .. SET` clause of both

- MySQL's `INSERT .. ON DUPLICATE KEY UPDATE SET`
- PostgreSQL's `INSERT .. ON CONFLICT .. DO UPDATE SET`

A natural syntax would be something like:

```java
ctx.insertInto(T)
   .columns(T.C1, T.C2, T.C3)
   .values(1, 2, 3)
   .values(4, 5, 6)
   .onDuplicateKeyUpdate()
   .setNonKeyToExcluded()
   .execute();
```

There are several reasonable options here:

- `setNonKeyToExcluded()` would set all the values that are not part of any key. This is interesting in MySQL, where `ON DUPLICATE KEY` isn't just about the primary key but about any of the keys
- `setNonPrimaryKeyToExcluded()` would set all the values that are not part of the primary key. This would allow for updating `UNIQUE` key values to a new value where this makes sense
- `setNonConflictingKeyToExcluded()` would set all the values that are not part of the `ON CONFLICT` key. This would be PostgreSQL specific and allow for updating all the values of columns that are not the `ON CONFLICT` key.

The fact that there are (at least) these 3 options hints at the feature request being more subtle than originally thought (as always with these "simple" things).

I'll do a bit more research. My intuition tells me that there should only be 1 API, and it would do the most "logical" thing depending on what clause had been called before `set()` (the MySQL or the PostgreSQL one).

**Comment by lukaseder** [comment]
In either case, I think the implementation here would optimally depend on a new feature for `EXCLUDED` support in general: https://github.com/jOOQ/jOOQ/issues/5214

While many workarounds exist (see https://github.com/jOOQ/jOOQ/issues/5214), a new query object model element should be added to allow for emulation of the feature:

- MySQL uses `VALUES(a)` instead of `EXCLUDED.a`
- Other dialects could have a `MERGE` specific implementation

Once https://github.com/jOOQ/jOOQ/issues/5214 is implemented, this feature request here would merely be convenience to generate such an `excluded(a)` expression for all relevant columns.

**Comment by lukaseder** [comment]
So, now that the infrastructure (`EXCLUDED` support as well as emulations) is there, let's tackle this. Again, my intuition here is not to provide too many similar APIs, given that users can always fall back to explicitly setting the `EXCLUDED` column set.

One thing seems certain: We only want those columns to be auto-`EXCLUDED`, which are part of the insert column list. E.g. given:

```sql
create table t (i int primary key, j int unique, k int, l int);
```

When writing:

```java
ctx.insertInto(T)
   .columns(T.I, T.J, T.K) // "Insert column list"
   .values(1, 2, 3)
   .onDuplicateKeyUpdate()
   .setExcluded()          // Name to be defined
   .execute();
```

Then, we want to include *at most* `(T.I, T.J, T.K)` in the `SET` clause, not, for example, `T.L`.

Now, what's specific for MySQL and this syntax is that it generates an `UPDATE` on conflicts both on `T.I` or `T.J`, so if there's a row `(0, 2, 100)`, the conflict is on `J`. It is very unlikely that in this case, users would want to update the primary key value to `I = 0`. At the same time, the primary key might be an identity in case there's a secondary key ("business key"), so it might be omitted from the insert column list.

If, however, the row is `(1, 100, 200)`, then the conflict on the primary key produces an update, and it is very likely that users want to update the unique key value to `J = 100`.

Things are different in PostgreSQL, where the `ON CONFLICT` clause requires a "conflict target", so users can specify explicitly if the interesting conflict happens on `I` or `J`. When the conflict target is `J`, then it is unlikely that the column `I` is included in the statement, e.g. because it is generated from an `IDENTITY`. If it isn't included, it won't be considered by the `SET EXCLUDED` clause, so we're fine, no primary key updates.

Given all the special cases and caveats, I think it is best for jOOQ not to be clever and opinionated here, jOOQ can't know the intention of the database design, so it can't include / exclude columns based on such assumptions. Even if a primary key update is very unlikely desired, it's the user's fault if they provide inconsistent data (differing primary key for identical secondary key). Hence, the best solution is to just set *all* the columns of the insert column list to their `EXCLUDED` value.

**Comment by lukaseder** [comment]
N.B: This is also consistent with H2's old `MERGE INTO` statement, which somewhat works similar to PostgreSQL's `INSERT .. ON CONFLICT`. By default (if omitted), its `KEY` clause references the primary key:

```sql
create table t (i int primary key, j int unique, k int, l int);
insert into t values (1, 1, 1, 1);
select * from final table (
  merge into t (i, j, k) values (1, 2, 3)
)
```

Produces (`J` updated to `2`, `L` left untouched)

```
|I  |J  |K  |L  |
|---|---|---|---|
|1  |2  |3  |1  |
```

Yet

```sql
delete from t;
insert into t values (1, 1, 1, 1);
select * from final table (
  merge into t (i, j, k) key(j) values (2, 1, 3)
)
```

Produces (primary key `I` updated, `L` left untouched)

```
|I  |J  |K  |L  |
|---|---|---|---|
|2  |1  |3  |1  |
```

**Comment by lukaseder** [comment]
I do wonder if the jOOQ 3.14 deprecation of the H2 style `MERGE` was premature: https://github.com/jOOQ/jOOQ/issues/10045

While it is poorly supported natively, it does add value in terms of simplicity.

**Comment by lukaseder** [comment]
Implemented for jOOQ 3.17.0

**Comment by lukaseder** [comment]
> I'll do a bit more research. My intuition tells me that there should only be 1 API, and it would do the most "logical" thing depending on what clause had been called before `set()` (the MySQL or the PostgreSQL one).

Well, the `setAllToExcluded()` method doesn't work for everyone, and the other types of methods will still be needed. I'll implement them here:

- https://github.com/jOOQ/jOOQ/issues/16522
